exonum-java-binding/core/src/main/java/com/exonum/binding/util/LibraryLoader.java

```diff
@@ -20,29 +20,81 @@
 import org.apache.logging.log4j.Logger;
 
 /**
- * A loader of the native shared library with Exonum framework bindings.
+ * A loader of the native shared library with Exonum framework bindings. It loads the native
+ * library and also verifies that it is compatible with the Java classes. The native library
+ * is compatible iff it has exactly the same version as this Java library. The revision
+ * from which they were built is currently not checked, but may be in the future (see ECR-3173),
+ * because the API between Java and native is considered internal and can be changed
+ * in an incompatible way in any revision.
  *
- * <p>To have library java_bindings available by its name,
- * add a path to the folder containing it to <code>java.library.path</code> property,
- * e.g.: <code>java -Djava.library.path=${EXONUM_HOME}/lib/native …</code>
+ * <p>To enable loading of java_bindings library, add a path to the folder containing it
+ * to <code>java.library.path</code> property, e.g.:
+ * <code>java -Djava.library.path=${EXONUM_HOME}/lib/native …</code>
+ *
+ * <p>This class is thread-safe.
+ *
+ * @see <a href="https://exonum.com/doc/version/0.11/get-started/java-binding/#installation">
+ *   Exonum Java installation instructions</a>
+ * @see <a href="https://exonum.com/doc/version/0.11/get-started/java-binding/#testing">
+ *   Build configuration to enable integration testing of Java services</a>
  */
 public final class LibraryLoader {
 
   private static final String BINDING_LIB_NAME = "java_bindings";
   private static final String JAVA_LIBRARY_PATH_PROPERTY = "java.library.path";
   private static final String DYNAMIC_LIBRARIES_ENV_VAR_WINDOWS = "PATH";
   private static final String DYNAMIC_LIBRARIES_ENV_VAR_UNIX = "LD_LIBRARY_PATH";
+  // TODO: Shall this class read the properties instead?
+  private static final String JAVA_BINDING_VERSION = "0.7.0-SNAPSHOT";
+  // TODO: Remove in ECR-3172
+  private static final boolean LIBRARY_VERSION_VERIFICATION_ENABLED = false;
 
   private static final Logger logger = LogManager.getLogger(LibraryLoader.class);
 
+  private static final LibraryLoader INSTANCE = new LibraryLoader(JAVA_BINDING_VERSION);
```

Comment:
How about to implement the singleton with lazy loading? It doesn't bring a big overhead but anyway.

Reply:
But it will never be actually "lazy" loaded, if I understand correctly — a class is loaded when its first static symbol is accessed (with some exception for compile-time constants, I believe), and the only publicly accessible symbol is `load`.

https://docs.oracle.com/javase/specs/jls/se11/html/jls-12.html#jls-12.4.1

Reply:
Agree, you are right. My doubt was that it'll be loaded in case no integration tests defined.